components/raftstore/src/store/fsm/store.rs

```diff
@@ -841,20 +841,52 @@ impl<EK: KvEngine, ER: RaftEngine, T: Transport> PollHandler<PeerFsm<EK, ER>, St
     }
 
     fn end(&mut self, peers: &mut [Option<impl DerefMut<Target = PeerFsm<EK, ER>>>]) {
-        let dur = if self.poll_ctx.has_ready {
+        if self.poll_ctx.has_ready {
             // Only enable the fail point when the store id is equal to 3, which is
             // the id of slow store in tests.
             fail_point!("on_raft_ready", self.poll_ctx.store_id() == 3, |_| {});
+        }
+        let mut latency_inspect = std::mem::take(&mut self.poll_ctx.pending_latency_inspect);
+        let mut dur = self.timer.saturating_elapsed();
 
-            if let Some(write_worker) = &mut self.poll_ctx.sync_write_worker {
+        if !latency_inspect.is_empty() {
+            for inspector in &mut latency_inspect {
+                inspector.record_store_process(dur);
+            }
+        }
+        let write_begin = TiInstant::now();
+        if let Some(write_worker) = &mut self.poll_ctx.sync_write_worker {
+            if self.poll_ctx.has_ready {
                 write_worker.write_to_db(false);
 
+                if !latency_inspect.is_empty() {
+                    for mut inspector in latency_inspect {
+                        inspector.record_store_write(write_begin.saturating_elapsed());
```

Comment:
Not sure if the original logic is on purpose, as the waterfall metrics may be accumulative? /cc @gengliqi @innerr